### DISCUSSION TITLE
Feature Request: Optional retries on HTTP requests

### DISCUSSION THREAD

**Discussion by DamianHeard** [XeroAPI/xero-python#80] [Ideas]
**Background**

`xero-python` uses `urllib3` under the hood. [By default, urllib3 will retry requests 3 times](https://urllib3.readthedocs.io/en/latest/user-guide.html#retrying-requests) which is not necessarily desirable, particularly if rate limiting is in play. In some case it might make sense to the API user to handle retries themselves.

**Proposed solution**

[`xero_python.api_client.configuration.Configuration`](https://github.com/XeroAPI/xero-python/blob/1c53f7d07b75e33621be39a9f85d888452056ee3/xero_python/api_client/configuration.py#L41)  should accept `retries` as an argument which in turn would be [passed down to the `urllib3.PoolManager` or `urllib3.ProxyManager` during initialisation](https://github.com/XeroAPI/xero-python/blob/1c53f7d07b75e33621be39a9f85d888452056ee3/xero_python/rest.py#L75).

I'm happy to put a PR up if you feel the change has merit.

**Comment by RettBehrens** [comment]
Hey @DamianHeard appreciate your feedback! I'm going to convert this to a discussion topic for the community to gauge temperature as it's not something we've seen requested previously. If you would like to put up a PR for the change we'd welcome the contribution!

**Comment by jaredthecoder** [comment]
@rdemarco-xero can we implement @DamianHeard 's suggestion above? This relates to the timeout issue and could fix that.

**Comment by timrichardson** [comment]
Waiting on a daily rate limit retry is not a very nice experience, I had to jump through hoops to manage the retry logic; the more conventional approach would be to pass our own retries object (@DamianHeard 's suggestion)
